### PR TITLE
Add keyedCache composable

### PR DESCRIPTION
**Pull request by davelopez**
Builds on top of #17145

Generic composable that reduces duplication in basic Pinia stores while maintaining type safety.

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by ElectronicBlueberry** [comment]
This looks like a useful composable! Two minor comments:

1. Could we use a different name? Maybe `useKeyedCache`. I do not like the current one, because it gives little indication over it's use, and it is not a store.

2. The composables parameters currently only take functions. I feel like they should both accept computed functions as well (can be easily casted to pure functions using `unref`), and the second one should also accept refs. Reason for this is that (especially the second parameter) would often be a getter / getter function, and in stores those should both be wrapped in computeds. Unwrapped functions are interpreted as actions by pinia, which getters should almost never be.

**Comment by davelopez** [comment]
Thanks for the feedback! I will make the changes :+1:

File: client/src/composables/keyedCache.test.ts [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+interface ItemData {
+    id: string;
+    name: string;
+}
+
+const fetchItem = jest.fn();
+const shouldFetch = jest.fn();
+
+describe("useSimpleKeyStore", () => {
```
Comment (ElectronicBlueberry) [inline client/src/composables/keyedCache.test.ts]:
```suggestion
describe("useKeyedCache", () => {
```

**Review by ElectronicBlueberry — APPROVED** [review]
Very nice abstraction!

File: client/src/composables/keyedCache.test.ts (hunk 1 above)
Comment (davelopez) [reply client/src/composables/keyedCache.test.ts]: Nice catch! Thanks!

**Comment by davelopez** [comment]
The selenium failures might be legit... investigating...

**Comment by davelopez** [comment]
> The selenium failures might be legit... investigating...

It was legit, it should be fixed now in https://github.com/galaxyproject/galaxy/pull/17148/commits/36fd9e3cd95b814aacf25e08b98114dfe8ed0985

**Review by dannon — APPROVED** [review]

**Pull request merged**